Cover: Add text color block support

aaronrobertshaw opened the pull request:
Depends on:
- https://github.com/WordPress/gutenberg/pull/41102

Related:
- https://github.com/WordPress/gutenberg/pull/38458
- https://github.com/WordPress/gutenberg/pull/40666

## What?

Adds text color support to the Cover block.

## Why?

This makes it easier to style the text color for a Cover's inner blocks as it can be set once only on the Cover block itself. It also allows us to clean up the Media & Text transforms so that text colors are cleanly passed between the resulting blocks.


## How?

- Opts into text color support
- Updates the Media & Text from/to transforms to handle text colors via block support attributes rather than updating inner blocks as well.
- Moves the light/dark adjustments of text color to the main block wrapper and applies them leveraging `:where` to allow for global styles or theme.json supplied colors to take precedence. (Whether this is desirable could be up for discussion)

## Testing Instructions

1. Create a couple of Media & Text blocks and assign them text and background colors, test both preset and custom colors.
2. Transform the Media & Text blocks to a Cover block
3. Check that the colors are assigned appropriately on the new Cover blocks i.e. the overlay is correct, the text color is only applied to the Cover block, and it all looks correct
4. Convert back to a Media & Text blocks and double check they appear as they did prior to the original transformation
5. Create a Cover block, add a gradient overlay and assign a text color.
6. Transform this Cover block to a Media & Text block and back again. It's color should be correct after each transform.
7. Try adjusting the overlay opacity for the Cover block between very low and very high values. The selected text color should be honored and not change.
8. Save the post and check the text color appears correctly on the front end
9. Back in the editor, remove the text color selection and again adjust the overlay opacity between the extremes.  The visual color of the text should be modified due to the light/dark handling of the Cover block.
10. Save again and make sure the light/dark adjusted color is reflected on the frontend.
11. Trial styling Cover block colors via theme.json and Global Styles. The global styles value should be applied over the light/dark defaults.

## Screenshots or screencast <!-- if applicable -->

| Adjusting Colors | Transforming Cover Blocks |
|---|---|
| <video src="https://user-images.githubusercontent.com/60436221/172309831-5a5cdb8b-baef-4bb8-8d8c-cd9b47bd2c77.mp4" /> | <video src="https://user-images.githubusercontent.com/60436221/172309841-5abd0959-6886-4c51-9e98-5c6d314d6dac.mp4" /> |

glendaviesnz commented:
Tested well for me. It looks like the unit test failures are related to the change though. Let me know if you want me to take a look at those, otherwise ping me when they are sorted and I will give it another review.

aaronrobertshaw committed:
Add text color block support to Cover block

aaronrobertshaw committed:
Prevent incorrect serialization of gradient class on wrapper

The Cover block previously (and likely accidentally) relied on gradient classes and styles not being serialized on the block wrapper as it didn't have text or background support. Its ad-hoc use of the same gradient attribute gives rise to the problem.

aaronrobertshaw commented:
> It looks like the unit test failures are related to the change though.

They were indeed. Although the problem runs a bit deeper than simply toggling on the text support as per this PR.

#### Problem Breakdown

The Cover block's ad-hoc use of the `gradient` attribute conflicts with the color block support. Actually, maybe it's more accurate to say that the color block support doesn't consider that the same attribute names might be reused on blocks that only partially adopt color block support.

Until the addition of text color support in this PR, the Cover block accidentally relied on the fact that it explicitly disabled text and background colors, which resulted in no gradient class being generated from that attribute.

Once either text or background support was enabled, the color block supports `addSaveProps` filter kicks in and within that, it only checks for any color support before then proceeding to generate the classes and styles. The inclusion of classes or styles at that point only considers whether the support attribute or style property is set and whether the support feature's serialization should be skipped.

Essentially, it assumes if a block is opting into a support feature, it won't have ad hoc use of an attribute that would normally be used for another of that support's features. An alternative example to the one in this PR might be, a block opting into background color support but implementing its own ad hoc use of the `textColor` block attribute.

#### Possible solutions

1. Update color block support to also check again for a feature's support before generating a class.
2. Skip the serialization of the gradient support for the Cover block

Option `#1` would be the one to fix the problem closes to its source. I erred away from this approach for now though as I'm not certain whether there'd be many other blocks that rely on the current behaviour without realising. This might be better addressed in a separate follow up.

#### Summary

Ultimately, I went with skipping the serialization of the gradient support despite the block not actually opting into it. It's a simple fix, won't require a deprecation to remove, and we can be sure it only impacts the Cover block for now, allowing exploration of Option `#1` in due course.

aaronrobertshaw commented:
@glendaviesnz the fixtures tests have passed locally and in the e2es now after tweaking the serialization of the gradient class. I think we can get this one reviewed with a view to landing it separately to any tightening of how color block support classes and styles are generated.

glendaviesnz reviewed (approved):
This tested well for me:

✅ Could transform between Cover and Media/Text blocks with preset and custom colors and settings were retained as expected both ways
✅ Text color set on Cover block was inherited by child blocks in editor and frontend
✅  Text color set in theme.json or global styles worked as expected in editor and frontend and could be overridden by user applied text color
✅  Auto assigning of light and dark text worked as expected unless a text color had been applied
✅ Applying of a gradient to the Cover block overlay still worked as expected

I agree that skip serialization is the best option for this PR and updated that color support can be followed up later.
